src/rules/LinearAlgebra/blas.jl

```diff
@@ -0,0 +1,148 @@
+module BLASRules
+
+using ..Enzyme
+using LinearAlgebra.BLAS
+
+const ConstOrDuplicated{T} = Union{Const{T},Duplicated{T}}
+const ConstOrBatchDuplicated{T} = Union{ConstOrDuplicated{T},BatchDuplicated{T}}
+
+_safe_similar(x::AbstractArray, n::Integer) = similar(x, n)
+_safe_similar(x::Ptr, n::Integer) = Array{eltype(x)}(undef, n)
+
+function _strided_tape(n::Integer, x::Union{AbstractArray,Ptr}, incx::Integer)
+    xtape = _safe_similar(x, n)
+    BLAS.blascopy!(n, x, incx, xtape, 1)
+    increment = 1
+    return xtape, increment
+end
+
+function _strided_range(n, x, incx)
+    r = range(1; step=abs(incx), length=n)
+    incx < 0 && return reverse(r)
+    return r
+end
+
+function _strided_view(n::Integer, x::AbstractArray, incx::Integer)
+    ind = _strided_range(n, x, incx)
+    return view(x, ind)
+end
+function _strided_view(n::Integer, x::Ptr, incx::Integer)
+    ind = _strided_range(n, x, incx)
+    dim = abs(last(ind) - first(ind)) + 1
+    y = Base.unsafe_wrap(Array, x, dim)
+    return view(y, ind)
+end
+
+function _maybe_primal_shadow(config, func, args)
+    needs_primal = EnzymeRules.needs_primal(config)
+    needs_shadow = EnzymeRules.needs_shadow(config)
+    if needs_primal || needs_shadow
+        r = func(args...)
+    else
+        r = nothing
+    end
+    primal = needs_primal ? r : nothing
+    shadow = needs_shadow ? zero(r) : nothing
+    return primal, shadow
+end
+
+_map_tuple(f, xs::Tuple...) = map(f, xs...)
+_map_tuple(f, xs...) = f(xs...)
+
+# axpy!(a, conj.(x), y)
+function _aconjxpy!(n, a, x, incx, y, incy)
+    xview = _strided_view(n, x, incx)
+    yview = _strided_view(n, y, incy)
+    yview .+= a .* conj.(xview)
+    return y
+end
+
+for (fname, Ttype) in ((:dot, :BlasReal), (:dotu, :BlasComplex), (:dotc, :BlasComplex))
+    @eval begin
+        function EnzymeRules.forward(
+            func::Const{typeof(BLAS.$fname)},
+            RT::Type{
+                <:Union{
+                    Const,DuplicatedNoNeed,Duplicated,BatchDuplicatedNoNeed,BatchDuplicated
+                },
+            },
+            n::Const{<:Integer},
+            X::ConstOrBatchDuplicated{<:Union{Ptr{T},AbstractArray{T}}},
+            incx::Const{<:Integer},
+            Y::ConstOrBatchDuplicated{<:Union{Ptr{T},AbstractArray{T}}},
+            incy::Const{<:Integer},
+        ) where {T<:BLAS.$Ttype}
+            RT <: Const && return func.val(n.val, X.val, incx.val, Y.val, incy.val)
```

Comment:
For some reason calling [the 2-arg `dot`](https://github.com/JuliaLang/julia/blob/7fc3e356e05ba46617c9b8fbbe228101eec0b6e6/stdlib/LinearAlgebra/src/blas.jl#L392), which forwards to the 5-arg `dot`, now segfaults with `Const` return type:
```suggestion
x, y, ∂x, ∂y = ntuple(_ -> randn(5), 4);
autodiff(Forward, Duplicated, Duplicated(x, ∂x), Duplicated(y, ∂y))  # fine
autodiff(Forward, Duplicated, Const(x), Duplicated(y, ∂y))  # fine
autodiff(Forward, Const, Duplicated(x, ∂x), Duplicated(y, ∂y))  # segfault
```
This is strange because only this line of code should be hit in this case, and all it does is call the primal function.

Reply:
The 2-arg methods also error for reverse-mode. These are the only remaining failures in the test suite.

Reply:
Open an issue?